src/buildercore/lifecycle.py

```diff
@@ -10,11 +10,74 @@
 import boto # route53 boto2 > route53 boto3
 from . import config, core
 from .core import boto_conn, find_ec2_instances, find_rds_instances, stack_all_ec2_nodes, current_ec2_node_id, NoPublicIps, NoRunningInstances
-from .utils import call_while, ensure
+from .utils import call_while, ensure, lmap
 from .context_handler import load_context, download_from_s3
 
 LOG = logging.getLogger(__name__)
 
+def _node_id(node):
+    name = core.tags2dict(node.tags)['Name']
+    nid = core.parse_stackname(name, all_bits=True, idx=True).get('cluster_id', 1)
+    return int(nid)
+
+def ec2_nodes(stackname):
+    "returns all non-terminated nodes associated with stackname, ordered by node-id"
+    nodes = find_ec2_instances(stackname, state='pending|running|shutting-down|stopping|stopped')
```

Comment:
if you are excluding `terminated`, I think also `shutting-down` should be excluded as a node like that is beyond being used anymore. It's a corner case

Reply:
(`shutting-down` is going towards being `terminated` without further intervention: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html)